File: JavaScript/0024-两两交换链表中的节点.js
class ListNode {
  constructor(val, next) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function swapPairs(head) {
  // 虚拟表头
  const data = new ListNode(0, head);
  let temp = data,
    cur = null,
    pre = null;

  while (temp.next !== null && temp.next.next !== null) {
    cur = temp.next.next;
    pre = temp.next;

    // 两两交换
    pre.next = cur.next;
    cur.next = pre;
    temp.next = cur;
    temp = pre;
  }

  return data.next;
}

File: Java/0024-两两交换链表中的节点.java
class ListNode {
  int val;
  ListNode next;

  ListNode() {
  }

  ListNode(int val) {
    this.val = val;
  }

  ListNode(int val, ListNode next) {
    this.val = val;
    this.next = next;
  }
}

class Solution {
  public ListNode swapPairs(ListNode head) {
    ListNode data = new ListNode(0, head);
    ListNode temp = data, cur, pre;
    while (temp.next != null && temp.next.next != null) {
      pre = temp.next;
      cur = temp.next.next;
      pre.next = cur.next;
      cur.next = pre;
      temp.next = cur;
      temp = pre;
    }
    return data.next;
  }
}
